python/tvm/relay/quantize/quantize.py

```diff
@@ -194,10 +258,239 @@ def annotate(graph):
+
+    def visit_func(expr):
+        if isinstance(expr, _expr.Call):
+            if expr.op == conv2d_op:
+                quant_weight = expr.args[1]
+                weight_data, weight_scale_var, _, _  = quant_weight.args
+                weight_scale = const_params[weight_scale_var].data
+
+                if expr.args[0].op != quantize_op and\
+                    expr.args[1].op == quantize_op:
+                    unified_scale = np.empty(weight_scale.asnumpy().shape, dtype='float32')
+                    # only weight shift possible
+                    parent = expr.args[0].args[0]
+                    assert parent.op == quantize_op
+                    _, data_scale_var, _, _, = parent.args
+                    data_scale = const_params[data_scale_var].data
+                    if data_scale.shape != weight_scale.shape:
+                        assert expr.args[0].op == _op.get("strided_slice")
+                        begin, end = slice_idx(expr.args[0].attrs.begin, expr.args[0].attrs.end)
+                        product = data_scale.asnumpy()[begin:end] * weight_scale.asnumpy()
+                    else:
+                        product = data_scale.asnumpy() * weight_scale.asnumpy()
+                    if mode == 'max':
+                        unified_scale = max(product)
+                    else:
+                        unified_scale = min(product)
+                    # (d * s_d) * (w * s_w) = (o * s_d * s_w)
+                    # (d * s_d) * (w * s_w') = (o * s_u)
+                    # s_w' = s_u/s_d
+                    gaps = unified_scale/product
+                    weight_scale_transform = np.empty(gaps.shape, dtype='float32')
+                    for i in range(0, gaps.shape[0]):
+                        shift_width = np.log2(gaps[i])
+                        if shift_width == 0:
+                            weight_scale_transform[i] = 1.0
+                        else:
+                            weight_scale_transform[i] = 2**shift_width
+                    new_weight_scale = weight_scale.asnumpy()*weight_scale_transform
+                    const_params[weight_scale_var] = _expr.const(new_weight_scale)
+                    return
+                elif expr.args[0].op == quantize_op and\
+                    expr.args[1].op != quantize_op:
+                    raise ValueError
+                elif expr.args[0].op != quantize_op and\
+                     expr.args[1].op != quantize_op:
+                    raise ValueError
+
+                quant_data = expr.args[0]
+                _, data_scale_var, _, _ = quant_data.args
+                data_scale = const_params[data_scale_var].data
+                assert len(data_scale.shape) == 1
+                assert len(weight_scale.shape) == 1
+                if data_scale.shape[0] == 1:
+                    assert weight_scale.shape[0] == 1
+                else:
+                    assert weight_scale.shape[0] == data_scale.shape[0] or\
```

Comment:
This assert is wrong under group convolution. reproduce by quantizing a ResNeXt.

Reply:
Thanks for the tip; can you point to an off-the-shelf definition of ResNeXT? I'm currently using gluon model zoo to easily import graphs and parameters; it would be great if something similar exists for ResNeXT.

Reply:
```
def load_gluon_cv():
  #block = gluoncv.model_zoo.resnet.resnet18_v1(pretrained=True)
  block = gluoncv.model_zoo.resnext.resnext50_32x4d(pretrained=True)
  #block = gluoncv.model_zoo.mobilenet.mobilenet_v2_1_0(pretrained=True)
  net, params = relay.frontend.from_mxnet(block,
                                          shape={"data": (1, 3, 224, 224)})
  return net, params
```

Reply:
Ok, I can reproduce this issue.
It seems handling general grouped convolutions are a little trickier than the pure depthwise case, because we currently view per-channel scales from the "input" perspective, but it should be doable.
Vanilla conv2d:
During calibration data and weight scales must be matched so that accumulation scales match.
During realize we can just multiply the scale tensors (vectors) together to get the result.
Depthwise conv2d:
During calibration data and weight scales do _not_ need to be matched because there is no accumulation across channels.
During realize we can just multiply the scales together and broadcasting will yield the correct result.
Grouped conv2d:
During calibration data and weight scales must be matched _locally_ within each group, but we can only manipulate the _data_ _scales_ to match the accumulation scale if we wish to avoid any sophisticated co-optimization between both data and weight scales.
During realize we cannot just broadcast using scale tensors themselves, but we might be able to do something like data_scale * weight_scale[0] -> downsample to get the output scale.

I will try this to see if it works.

Reply:
I have a temp solution running https://github.com/antinucleon/tvm/commit/2035899a9b57b24bda16eb40a37fae66cd1bb55e

The ResNext will drop 1% (which I am not too worried yet because once full pipeline is able to run, we can tune it later.

Reply:
Ok, I have something similar implemented and I am evaluating the current accuracy. Looks like I get 76.4% for per-channel with 64 calibration samples and per-layer will be > 77% (still running).